node/src/components/consensus/era_supervisor/era.rs

```diff
@@ -195,11 +190,19 @@ impl DataSize for Era {
                 } else {
                     (*highway).estimate_heap_size()
                 }
+            } else if let Some(zug) = any_ref.downcast_ref::<Zug<ClContext>>() {
+                if *CASPER_ENABLE_DETAILED_CONSENSUS_METRICS {
+                    let detailed = (*zug).estimate_detailed_heap_size();
+                    match serde_json::to_string(&detailed) {
+                        Ok(encoded) => debug!(%encoded, "consensus memory metrics"),
+                        Err(err) => warn!(%err, "error encoding consensus memory metrics"),
+                    }
+                    detailed.total()
+                } else {
+                    (*zug).estimate_heap_size()
+                }
```

Comment:
Would it be possible to downcast to something like `&dyn DataSize` regardless of which consensus that is?

Reply:
Unfortunately `DataSize` isn't object safe because of its associated `const`s.